### ISSUE TITLE
docker build, invalid environment variable:

### ISSUE THREAD

**Issue by firjest** [moby/moby#38439]
**Description**

old image may have empty environment, and the build from this image, will  returns message:
 invalid environment variable:
**Steps to reproduce the issue:**
1.old image has environment like:
"Env": [
"",
"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
"container=docker"
],

and the code is return error:
opts/opts.go
func ValidateEnv(val string) (string, error) {
arr := strings.Split(val, "=")
if arr[0] == "" {
return "", fmt.Errorf("invalid environment variable: %s", val)
}
2.
3.

**Describe the results you received:**
 invalid environment variable:

**Describe the results you expected:**
did we need to support such old image?
we can filter empty variable to do that

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
(paste your output here)
```

**Output of `docker info`:**

```
(paste your output here)
```

**Additional environment details (AWS, VirtualBox, physical, [etc.):**] 
 
pre issue is closed: (https://github.com/moby/moby/issues/30356)

**Comment by thaJeztah** [comment]
@firjest do you know how that old image was built; is that an image you built yourself? And if so; what version of docker was used to build that image?

**Comment by thaJeztah** [comment]
Thanks for the extra info; looking at that example, I think the validation is "ok", but very specific about an empty string, and, in the current version still allows for empty environment variables to be set (if they're white-space only);


```bash
docker run --name bla -e "   " busybox

docker inspect --format '{{json .Config.Env}}' bla
["   ","PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"]


docker commit bla repro-38439
sha256:e5967e465ebd5707e6aa39356d410e617bdd64739eaf0ee8ca80ca1837680349

docker inspect --format '{{json .Config.Env}}' repro-38439
["   ","PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"]


docker build -t repro-38439-2 -<<'EOF'
FROM repro-38439
ENV test=bla
EOF


docker inspect --format '{{json .Config.Env}}' repro-38439-2
["   ","PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin","test=bla"]
```

Looks like this also is accepted;

```bash
docker run --name bla -e "  =  " busybox
docker inspect --format '{{json .Config.Env}}' bla
["  =  ","PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"]
```


Thinking out loud, I think we should; validate after trimming whitespace

And either be _more_ strict (fail on any whitespace-only env-var), or instead of failing, print a warning, and remove the invalid env-var.

FWIW, we removed other validation for environment variables (and leave validation to the container's process itself), but I think it should be safe to warn/ignore empty env-vars.


I'll label this with "bug" and "enhancement" (as it can be seen as either one, depending how you look at it 😅 )


@tonistiigi @tiborvass @vdemeester WDYT?

**Comment by firjest** [comment]
> looking at that example, I think the validation is "ok"

It is true that the verification is valid, but if you really encounter a similar image, do we need to be compatible?

**Comment by vdemeester** [comment]
> And either be more strict (fail on any whitespace-only env-var), or instead of failing, print a warning, and remove the invalid env-var.

Indeed, and I would vote for a warning (i.e. no hard fail on those empty vars)

**Comment by thaJeztah** [comment]
I pushed a work-in-progress pull request (was working on it this afternoon, but found that it's not covering all cases yet; decided to push it early to get more eyes on the proposed changes); https://github.com/moby/moby/pull/38449
